### PR TITLE
Add `span`

### PR DESCRIPTION
**Pull request by elbeno**
Adding span required `to_address` and `type_identity`, both fairly small additions.
Also fixed up some C++17 quality issues.

### CODE REVIEW

File: include/stdx/span.hpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        return *std::prev(end());
+    }
+
+    [[nodiscard]] constexpr auto operator[](size_type idx) const -> reference {
+        return data()[idx];
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto first() const -> span<element_type, Count> {
+        return span<element_type, Count>{ptr, Count};
```
Comment (mjcaisse-intel) [inline include/stdx/span.hpp:176]: **suggestion**: ensure that `Count` is valid when the `Extent` is not dynamic.

File: include/stdx/span.hpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    [[nodiscard]] constexpr auto first(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr, count};
+    }
+
+    template <std::size_t Count>
+    [[nodiscard]] constexpr auto last() const -> span<element_type, Count> {
```
Comment (mjcaisse-intel) [inline include/stdx/span.hpp:184]: **suggestion**: ensure that `Count` is valid when the `Extent` is not dynamic.

File: include/stdx/span.hpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    [[nodiscard]] constexpr auto last(size_type count) const
+        -> span<element_type, dynamic_extent> {
+        return {ptr + this->size() - count, count};
+    }
+
+    template <std::size_t Offset, std::size_t Count = dynamic_extent>
+    [[nodiscard]] constexpr auto subspan() const {
+        if constexpr (Count != dynamic_extent) {
+            return span<element_type, Count>{ptr + Offset, Count};
```
Comment (mjcaisse-intel) [inline include/stdx/span.hpp:195]: **suggestion**: ensure that `Count` and `Offset` form a valid pair when the `Extent` is not dynamic.

File: include/stdx/span.hpp [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        } else if constexpr (Extent != dynamic_extent) {
+            return span<element_type, Extent - Offset>{ptr + Offset,
```
Comment (mjcaisse-intel) [inline include/stdx/span.hpp:197]: **suggestion**: ensure that `Offset <= Extent`.

File: include/stdx/span.hpp [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    }
+};
+
+// NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast)
+template <class T, std::size_t N> auto as_bytes(span<T, N> s) noexcept {
+    if constexpr (N == dynamic_extent) {
+        return span{reinterpret_cast<std::byte const *>(s.data()),
+                    s.size_bytes()};
+    } else {
+        constexpr auto Size = sizeof(T) * N;
```
Comment (mjcaisse-intel) [inline include/stdx/span.hpp:218]: **question**: `size_bytes` should be `constexpr` when `N != dynamic_extent`, no?

**Review by mjcaisse-intel — CHANGES_REQUESTED** [review]
This is really great! `span` is a favorite tool for me. Great to see it coming to `stdx`. Some minor suggestions and questions.

File: include/stdx/span.hpp (hunk 5 above)
Comment (elbeno) [reply include/stdx/span.hpp]: Yes - that's simpler. Changed.

**Review by mjcaisse-intel — APPROVED** [review]
Excited to see `span` join stdx!

**Pull request merged**